microcosm/opaque.py

```diff
@@ -0,0 +1,86 @@
+from contextdecorator import ContextDecorator
+from collections import MutableMapping
+from copy import deepcopy
+from types import MethodType
+
+from microcosm.api import binding
+
+
+class Opaque(MutableMapping):
+    """
+    The Opaque collaborator and its assocaited context manager are extremely useful for
+    instantiating context specific data during transaction processing. Example:
+
+        from microcosm.opaque import Opaque
+
+        # set up
+        def special_opaque_data(some, stuff):
+            return "{} {}".format(some, stuff)
+        some = "some"
+        stuff = "stuff"
+
+        opaque = Opaque()
+
+        # usage
+        print opaque.opaque_data_func()
+
+        with opaque.opaque_data(special_opaque_data, some, stuff):
+            print opaque.opaque_data_func()
+
+        print opaque.opaque_data_func()
+
+    Outputs:
+
+        {}
+        some stuff
+        {}
+
+
+    Note: opaque.opaque_data may also be used as a decorator.
+
+    """
+    def __init__(self, *args, **kwargs):
+        self._store = dict()
+        self.update(dict(*args, **kwargs))
+        self.bind = _make_bind(self)
+
+    def __getitem__(self, key):
+        return self._store[key]
+
+    def __setitem__(self, key, value):
+        self._store[key] = value
+
+    def __delitem__(self, key):
+        del self._store[key]
+
+    def __iter__(self):
+        return iter(self._store)
+
+    def __len__(self):
+        return len(self._store)
+
+    def as_dict(self):
+        return self._store
+
+
+def _make_bind(opaque):
+    class OpaqueData(ContextDecorator):
+        def __init__(self, opaque_data_func, *args, **kwargs):
+
+            def context_specific_data_func(self):
+                return opaque_data_func(*args, **kwargs)
+
+            self.context_specific_data_func = MethodType(context_specific_data_func, self)
+
+        def __enter__(self):
+            self.original_store = deepcopy(opaque._store)
```

Comment:
I originally did this in the `__init__` but that runs the first time a function is decorated and so it will capture the value of when the function being decorated is decorated, not the state of `opaque._store` before the function being decorated is actually called.